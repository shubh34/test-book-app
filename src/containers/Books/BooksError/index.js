import PageHeader from '../../../components/PageHeader';

const BooksError = () => {
  return (
    <div
      style={{
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        minHeight: '600px'
      }}
    >
      <PageHeader header="Something Gone Wrong!!! Please try later"> </PageHeader>
    </div>
  );
};
export default BooksError;
