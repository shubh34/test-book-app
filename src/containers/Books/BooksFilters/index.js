import Button from '../../../components/Button';
import { booksPerPageOptions } from '../../../configs';
import { useSelector } from 'react-redux';
import { getActivePerPageCount } from '../../../states/books/selectors';
import { useLocation, useHistory } from 'react-router';
import routes from '../../../configs/routes';
// import BooksSearch from '../BooksSearch';

const BooksFilters = () => {
  const location = useLocation();
  const history = useHistory();
  const activePerPageCount = useSelector(getActivePerPageCount);
  const searchParams = new URLSearchParams(location.search);
  const handleActivePageCountChange = (event) => {
    searchParams.set('per_page', event.target.id);
    history.push({
      path: routes.BOOKS.resolve(),
      search: `?per_page=${event.target.id}`
    });
  };
  return (
    <div style={{ display: 'flex', alignItems: 'center', padding: '16px 0' }}>
      Show
      {booksPerPageOptions.map((option) => (
        <Button
          key={option}
          isActive={option === activePerPageCount}
          id={option}
          name={option}
          onClick={handleActivePageCountChange}
        />
      ))}
      Entries
    </div>
  );
};
export default BooksFilters;
