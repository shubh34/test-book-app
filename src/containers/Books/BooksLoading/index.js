import { useSelector } from 'react-redux';
import { getActivePerPageCount } from '../../../states/books/selectors';
import Table from '../../../components/Table';
import { bookTableConfig } from '../BooksList/index.config';

const BooksLoading = () => {
  const pageCount = useSelector(getActivePerPageCount);
  return (
    <>
      {[...Array(pageCount)].map((count, index) => (
        <Table.Row key={index} isGhost>
          {bookTableConfig.map((tableHeader) => (
            <Table.Cell
              key={`row-${index}-${tableHeader.id}`}
              title={tableHeader.mobile_name}
              styling={{ ...tableHeader.styling, height: '30px' }}
            />
          ))}
        </Table.Row>
      ))}
    </>
  );
};
export default BooksLoading;
