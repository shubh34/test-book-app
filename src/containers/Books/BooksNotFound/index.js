import PageHeader from '../../../components/PageHeader';

const BooksNotFound = () => {
  return (
    <div
      style={{
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        minHeight: '600px'
      }}
    >
      <PageHeader> No Books found!! </PageHeader>
    </div>
  );
};
export default BooksNotFound;
