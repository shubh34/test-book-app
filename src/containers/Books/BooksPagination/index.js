import Button from '../../../components/Button';
import {
  getActivePageNumber,
  getActivePageStart,
  shouldShowPreviousButton,
  shouldShowNextButton,
  shouldShowPagination
} from '../../../states/books/selectors';
import routes from '../../../configs/routes';
import { paginationButtonsToShow } from '../../../configs';
import { Component } from 'react';
import { connect } from 'react-redux';

const mapStateToProps = (state) => ({
  activePageNumber: getActivePageNumber(state),
  activePageStart: getActivePageStart(state),
  showPagination: shouldShowPagination(state),
  showPrevious: shouldShowPreviousButton(state),
  showNext: shouldShowNextButton(state)
});

const mapDispatch = (dispatch) => ({});

class BooksPagination extends Component {
  constructor(props) {
    super(props);
    this.handlePageNumberClick = this.handlePageNumberClick.bind(this);
    this.handlePageNext = this.handlePageNext.bind(this);
    this.handlePagePrevious = this.handlePagePrevious.bind(this);
  }
  handlePageNumberClick(event) {
    const { searchParams, history } = this.props;
    searchParams.set('page', event.target.id);
    history.push({
      path: routes.BOOKS.resolve(),
      search: `?${searchParams.toString()}`
    });
  }
  handlePageNext() {
    const { searchParams, history, activePageNumber } = this.props;
    searchParams.set('page', activePageNumber + 1);
    history.push({
      path: routes.BOOKS.resolve(),
      search: `?${searchParams.toString()}`
    });
  }
  handlePagePrevious() {
    const { searchParams, history, activePageNumber } = this.props;
    searchParams.set('page', activePageNumber - 1);
    history.push({
      path: routes.BOOKS.resolve(),
      search: `?${searchParams.toString()}`
    });
  }
  render() {
    const {
      showPagination,
      showPrevious,
      activePageStart,
      activePageNumber,
      showNext
    } = this.props;
    if (!showPagination) {
      return null;
    }
    return (
      <div style={{ display: 'flex', padding: '16px 0' }}>
        <Button name="Previous" disabled={!showPrevious} onClick={this.handlePagePrevious} />
        {[...Array(paginationButtonsToShow)].map((count, index) => {
          const buttonNumber = activePageStart + index;
          return (
            <Button
              key={`page-${buttonNumber}`}
              name={buttonNumber}
              id={buttonNumber}
              isActive={buttonNumber === activePageNumber}
              onClick={this.handlePageNumberClick}
            />
          );
        })}
        <Button name={'Next'} disabled={!showNext} onClick={this.handlePageNext} />
      </div>
    );
  }
}
export default connect(mapStateToProps, mapDispatch)(BooksPagination);
