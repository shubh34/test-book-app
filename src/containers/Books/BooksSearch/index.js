
const BooksSearch = () => {
  return (
    <label>
      Search: <input type="search" />
    </label>
  );
};
export default BooksSearch;
