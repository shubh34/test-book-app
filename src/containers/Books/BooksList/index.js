import { useSelector } from 'react-redux';
import { getBooks, fetchBookInProgress } from '../../../states/books/selectors';
import { bookTableConfig } from './index.config';
import Table from '../../../components/Table';
import BooksLoading from '../BooksLoading';

const BooksList = () => {
  const books = useSelector(getBooks);
  const isApiCallInProgress = useSelector(fetchBookInProgress);
  return (
    <Table.Container>
      <Table.Header>
        <Table.Row>
          {bookTableConfig.map((tableHeader) => (
            <Table.HeaderCell key={`header-${tableHeader.id}`} style={tableHeader.style}>
              {tableHeader.name}
            </Table.HeaderCell>
          ))}
        </Table.Row>
      </Table.Header>
      <Table.Body>
        {isApiCallInProgress && <BooksLoading />}
        {!isApiCallInProgress &&
          books.map((book) => (
            <Table.Row key={book.id}>
              {bookTableConfig.map((tableHeader) => (
                <Table.Cell
                  key={`row-${book.id}-${tableHeader.id}`}
                  title={tableHeader.mobile_name}
                  styling={tableHeader.styling}
                >
                  {book[tableHeader.path]}
                </Table.Cell>
              ))}
            </Table.Row>
          ))}
      </Table.Body>
    </Table.Container>
  );
};
export default BooksList;
