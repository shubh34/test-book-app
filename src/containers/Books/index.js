import { useDispatch } from 'react-redux';
import { useEffect } from 'react';
import { fetchBooks } from '../../states/books/actions';
import { useBooks } from './useBooks';
import BooksList from './BooksList';
import BooksFilters from './BooksFilters';
import BooksPagination from './BooksPagination';
import Container from '../../components/Container';
import PageHeader from '../../components/PageHeader';

const Books = () => {
  const dispatch = useDispatch();
  const { page, filters, itemsPerPage } = useBooks();
  const stringifiedFilters = JSON.stringify(filters);
  useEffect(() => {
    const filters = JSON.parse(stringifiedFilters);
    window.scrollTo(0, 0);
    dispatch(fetchBooks({ page, filters, itemsPerPage }));
  }, [dispatch, page, stringifiedFilters, itemsPerPage]);

  return (
    <Container>
      <PageHeader header={'Find My Books'} />
      <BooksFilters />
      <BooksList />
      <BooksPagination />
    </Container>
  );
};
export default Books;
