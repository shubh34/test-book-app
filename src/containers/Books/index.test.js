import { setUpConnectedComponent } from '../../../test/utils';

import Books from './';

describe('Books', () => {
  it('should render Books', () => {
    const hasThunk = true;
    const { container } = setUpConnectedComponent({
      Component: Books,
      props: {},
      stateKey: 'DEFAULT_WITH_BOOKS',
      hasThunk
    });
    expect(container).toMatchSnapshot();
  });
  it('should render Books in Loading State', () => {
    const hasThunk = true;
    const { container } = setUpConnectedComponent({
      Component: Books,
      props: {},
      stateKey: 'BOOKS_WITH_API_IN_PROGRESS',
      hasThunk
    });
    expect(container).toMatchSnapshot();
  });
  it('should render Books Not Found', () => {
    const hasThunk = true;
    const { container } = setUpConnectedComponent({
      Component: Books,
      props: {},
      stateKey: 'DEFAULT',
      hasThunk
    });
    expect(container).toMatchSnapshot();
  });
  it('should render Books Not Found when error', () => {
    const hasThunk = true;
    const { container } = setUpConnectedComponent({
      Component: Books,
      props: {},
      stateKey: 'BOOKS_WITH_API_FAILURE',
      hasThunk
    });
    expect(container).toMatchSnapshot();
  });
});
