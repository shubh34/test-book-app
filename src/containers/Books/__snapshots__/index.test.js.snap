// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Books should render Books 1`] = `
<div>
  <div
    class="sc-iBPRYJ gokKvP"
  >
    <h3
      class="sc-jSgupP lnXuzl"
    >
      Find My Books
    </h3>
    <div
      style="display: flex; align-items: center; padding: 16px 0px;"
    >
      Show
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="10"
        id="10"
        name="10"
      >
        10
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="20"
        id="20"
        name="20"
      >
        20
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="40"
        id="40"
        name="40"
      >
        40
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="80"
        id="80"
        name="80"
      >
        80
      </button>
      Entries
    </div>
    <table
      class="sc-bdfBwQ SZIUD"
    >
      <thead
        class="sc-gsTCUz eVsNgG"
      >
        <tr
          class="sc-dlfnbm fhHUUt"
        >
          <th
            class="sc-hKgILt eemWPu"
          >
            Name
          </th>
          <th
            class="sc-hKgILt eemWPu"
          >
            Authors
          </th>
          <th
            class="sc-hKgILt eemWPu"
          >
            Publication Year
          </th>
          <th
            class="sc-hKgILt eemWPu"
          >
            Publication Location
          </th>
          <th
            class="sc-hKgILt eemWPu"
          >
            Pages
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="sc-dlfnbm fhHUUt"
        >
          <td
            class="sc-eCssSg hMVDoG"
            title="Name"
          >
            Book Title
          </td>
          <td
            class="sc-eCssSg jMTdNr"
            title="Authors"
          >
            Author 1,Author 2
          </td>
          <td
            class="sc-eCssSg cWxcEj"
            title="Year"
          >
            2020
          </td>
          <td
            class="sc-eCssSg kkFOlx"
            title="Location"
          >
            City, Country
          </td>
          <td
            class="sc-eCssSg fPEa-ds"
            title="Pages"
          >
            104
          </td>
        </tr>
      </tbody>
    </table>
    <div
      style="display: flex; padding: 16px 0px;"
    >
      <button
        class="sc-gKsewC jyHuJi"
        disabled=""
        name="Previous"
      >
        Previous
      </button>
      <button
        class="sc-gKsewC fzYsXG"
        data-testid="1"
        id="1"
        name="1"
      >
        1
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="2"
        id="2"
        name="2"
      >
        2
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="3"
        id="3"
        name="3"
      >
        3
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="4"
        id="4"
        name="4"
      >
        4
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="5"
        id="5"
        name="5"
      >
        5
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="6"
        id="6"
        name="6"
      >
        6
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        name="Next"
      >
        Next
      </button>
    </div>
  </div>
</div>
`;

exports[`Books should render Books Not Found 1`] = `
<div>
  <div
    class="sc-iBPRYJ gokKvP"
  >
    <h3
      class="sc-jSgupP lnXuzl"
    >
      Find My Books
    </h3>
    <div
      style="display: flex; align-items: center; padding: 16px 0px;"
    >
      Show
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="10"
        id="10"
        name="10"
      >
        10
      </button>
      <button
        class="sc-gKsewC fzYsXG"
        data-testid="20"
        id="20"
        name="20"
      >
        20
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="40"
        id="40"
        name="40"
      >
        40
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="80"
        id="80"
        name="80"
      >
        80
      </button>
      Entries
    </div>
    <table
      class="sc-bdfBwQ SZIUD"
    >
      <thead
        class="sc-gsTCUz eVsNgG"
      >
        <tr
          class="sc-dlfnbm fhHUUt"
        >
          <th
            class="sc-hKgILt eemWPu"
          >
            Name
          </th>
          <th
            class="sc-hKgILt eemWPu"
          >
            Authors
          </th>
          <th
            class="sc-hKgILt eemWPu"
          >
            Publication Year
          </th>
          <th
            class="sc-hKgILt eemWPu"
          >
            Publication Location
          </th>
          <th
            class="sc-hKgILt eemWPu"
          >
            Pages
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <div
      style="display: flex; justify-content: center; align-items: center; min-height: 600px;"
    >
      <h3
        class="sc-jSgupP lnXuzl"
      >
        No Books found!!
      </h3>
    </div>
  </div>
</div>
`;

exports[`Books should render Books Not Found when error 1`] = `
<div>
  <div
    class="sc-iBPRYJ gokKvP"
  >
    <h3
      class="sc-jSgupP lnXuzl"
    >
      Find My Books
    </h3>
    <div
      style="display: flex; align-items: center; padding: 16px 0px;"
    >
      Show
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="10"
        id="10"
        name="10"
      >
        10
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="20"
        id="20"
        name="20"
      >
        20
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="40"
        id="40"
        name="40"
      >
        40
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="80"
        id="80"
        name="80"
      >
        80
      </button>
      Entries
    </div>
    <table
      class="sc-bdfBwQ SZIUD"
    >
      <thead
        class="sc-gsTCUz eVsNgG"
      >
        <tr
          class="sc-dlfnbm fhHUUt"
        >
          <th
            class="sc-hKgILt eemWPu"
          >
            Name
          </th>
          <th
            class="sc-hKgILt eemWPu"
          >
            Authors
          </th>
          <th
            class="sc-hKgILt eemWPu"
          >
            Publication Year
          </th>
          <th
            class="sc-hKgILt eemWPu"
          >
            Publication Location
          </th>
          <th
            class="sc-hKgILt eemWPu"
          >
            Pages
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="sc-dlfnbm bqEDSD"
        >
          <td
            class="sc-eCssSg gGGlRr"
            title="Name"
          />
          <td
            class="sc-eCssSg erUswG"
            title="Authors"
          />
          <td
            class="sc-eCssSg jhCely"
            title="Year"
          />
          <td
            class="sc-eCssSg ShpYU"
            title="Location"
          />
          <td
            class="sc-eCssSg bkMOxV"
            title="Pages"
          />
        </tr>
        <tr
          class="sc-dlfnbm bqEDSD"
        >
          <td
            class="sc-eCssSg gGGlRr"
            title="Name"
          />
          <td
            class="sc-eCssSg erUswG"
            title="Authors"
          />
          <td
            class="sc-eCssSg jhCely"
            title="Year"
          />
          <td
            class="sc-eCssSg ShpYU"
            title="Location"
          />
          <td
            class="sc-eCssSg bkMOxV"
            title="Pages"
          />
        </tr>
      </tbody>
    </table>
    <div
      style="display: flex; justify-content: center; align-items: center; min-height: 600px;"
    >
      <h3
        class="sc-jSgupP lnXuzl"
      >
        Something Gone Wrong!!! Please try later
      </h3>
    </div>
  </div>
</div>
`;

exports[`Books should render Books in Loading State 1`] = `
<div>
  <div
    class="sc-iBPRYJ gokKvP"
  >
    <h3
      class="sc-jSgupP lnXuzl"
    >
      Find My Books
    </h3>
    <div
      style="display: flex; align-items: center; padding: 16px 0px;"
    >
      Show
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="10"
        id="10"
        name="10"
      >
        10
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="20"
        id="20"
        name="20"
      >
        20
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="40"
        id="40"
        name="40"
      >
        40
      </button>
      <button
        class="sc-gKsewC jyHuJi"
        data-testid="80"
        id="80"
        name="80"
      >
        80
      </button>
      Entries
    </div>
    <table
      class="sc-bdfBwQ SZIUD"
    >
      <thead
        class="sc-gsTCUz eVsNgG"
      >
        <tr
          class="sc-dlfnbm fhHUUt"
        >
          <th
            class="sc-hKgILt eemWPu"
          >
            Name
          </th>
          <th
            class="sc-hKgILt eemWPu"
          >
            Authors
          </th>
          <th
            class="sc-hKgILt eemWPu"
          >
            Publication Year
          </th>
          <th
            class="sc-hKgILt eemWPu"
          >
            Publication Location
          </th>
          <th
            class="sc-hKgILt eemWPu"
          >
            Pages
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="sc-dlfnbm bqEDSD"
        >
          <td
            class="sc-eCssSg gGGlRr"
            title="Name"
          />
          <td
            class="sc-eCssSg erUswG"
            title="Authors"
          />
          <td
            class="sc-eCssSg jhCely"
            title="Year"
          />
          <td
            class="sc-eCssSg ShpYU"
            title="Location"
          />
          <td
            class="sc-eCssSg bkMOxV"
            title="Pages"
          />
        </tr>
        <tr
          class="sc-dlfnbm bqEDSD"
        >
          <td
            class="sc-eCssSg gGGlRr"
            title="Name"
          />
          <td
            class="sc-eCssSg erUswG"
            title="Authors"
          />
          <td
            class="sc-eCssSg jhCely"
            title="Year"
          />
          <td
            class="sc-eCssSg ShpYU"
            title="Location"
          />
          <td
            class="sc-eCssSg bkMOxV"
            title="Pages"
          />
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
